File: C++/ScreenLockingPatterns.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

const int GRID_SIZE = 3;
const int NUM_POINTS = GRID_SIZE * GRID_SIZE;

int getIndex(char point) {
    return point - 'A';
}

bool isValidMove(int from, int to, const std::vector<bool>& visited, const std::vector<std::vector<int>>& skip) {
    if (visited[to]) {
        return false;
    }
    if (skip[from][to] != -1 && !visited[skip[from][to]]) {
        return false;
    }
    return true;
}

int dfs(int current, int remainingLength, std::vector<bool>& visited, const std::vector<std::vector<int>>& skip) {
    if (remainingLength == 0) {
        return 1;
    }

    visited[current] = true;
    int count = 0;

    for (int i = 0; i < NUM_POINTS; ++i) {
        if (isValidMove(current, i, visited, skip)) {
            count += dfs(i, remainingLength - 1, visited, skip);
        }
    }

    visited[current] = false;
    return count;
}

unsigned int countPatternsFrom(char firstDot, unsigned short length) {
    if (length <= 0 || length > NUM_POINTS) {
        return 0;
    }

    std::vector<std::vector<int>> skip(NUM_POINTS, std::vector<int>(NUM_POINTS, -1));

    skip[getIndex('A')][getIndex('C')] = skip[getIndex('C')][getIndex('A')] = getIndex('B');
    skip[getIndex('A')][getIndex('I')] = skip[getIndex('I')][getIndex('A')] = getIndex('E');
    skip[getIndex('A')][getIndex('G')] = skip[getIndex('G')][getIndex('A')] = getIndex('D');
    skip[getIndex('B')][getIndex('H')] = skip[getIndex('H')][getIndex('B')] = getIndex('E');
    skip[getIndex('C')][getIndex('I')] = skip[getIndex('I')][getIndex('C')] = getIndex('F');
    skip[getIndex('C')][getIndex('G')] = skip[getIndex('G')][getIndex('C')] = getIndex('E');
    skip[getIndex('D')][getIndex('F')] = skip[getIndex('F')][getIndex('D')] = getIndex('E');
    skip[getIndex('G')][getIndex('I')] = skip[getIndex('I')][getIndex('G')] = getIndex('H');

    std::vector<bool> visited(NUM_POINTS, false);
    return dfs(getIndex(firstDot), length - 1, visited, skip);
}

int main() {
    std::cout << countPatternsFrom('D', 3) << std::endl; // Expected output: 37
    return 0;
}

File: Screen_Locking_Patterns.py
GRID_SIZE = 3
NUM_POINTS = GRID_SIZE * GRID_SIZE

def get_index(point):
    return ord(point) - ord('A')

def is_valid_move(from_idx, to_idx, visited, skip):
    if visited[to_idx]:
        return False
    if skip[from_idx][to_idx] != -1 and not visited[skip[from_idx][to_idx]]:
        return False
    return True

def dfs(current, remaining_length, visited, skip):
    if remaining_length == 0:
        return 1

    visited[current] = True
    count = 0

    for i in range(NUM_POINTS):
        if is_valid_move(current, i, visited, skip):
            count += dfs(i, remaining_length - 1, visited, skip)

    visited[current] = False
    return count

def count_patterns_from(first_point, length):
    if length <= 0 or length > NUM_POINTS:
        return 0

    skip = [[-1] * NUM_POINTS for _ in range(NUM_POINTS)]

    skip[get_index('A')][get_index('C')] = skip[get_index('C')][get_index('A')] = get_index('B')
    skip[get_index('A')][get_index('I')] = skip[get_index('I')][get_index('A')] = get_index('E')
    skip[get_index('A')][get_index('G')] = skip[get_index('G')][get_index('A')] = get_index('D')
    skip[get_index('B')][get_index('H')] = skip[get_index('H')][get_index('B')] = get_index('E')
    skip[get_index('C')][get_index('I')] = skip[get_index('I')][get_index('C')] = get_index('F')
    skip[get_index('C')][get_index('G')] = skip[get_index('G')][get_index('C')] = get_index('E')
    skip[get_index('D')][get_index('F')] = skip[get_index('F')][get_index('D')] = get_index('E')
    skip[get_index('G')][get_index('I')] = skip[get_index('I')][get_index('G')] = get_index('H')

    visited = [False] * NUM_POINTS
    return dfs(get_index(first_point), length - 1, visited, skip)

# Testing the function
print(count_patterns_from('D', 3))  # Expected output: 37
